evm/chains/sharding/mainchain_handler/vmc_handler.py

```diff
@@ -0,0 +1,219 @@
+    GASPRICE,
+    TX_GAS,
+)
+
+from evm.chains.sharding.mainchain_handler.vmc_utils import (
+    decode_vmc_call_result,
+    get_valmgr_abi,
+    get_valmgr_addr,
+    get_valmgr_bytecode,
+    get_valmgr_code,
+    get_valmgr_sender_addr,
+    mk_vmc_tx_obj,
+)
+
+class VMCHandler:
+
+    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMCHandler")
+
+    def __init__(self, chain_handler, primary_addr):
+        """
+        :param primary_addr: address in bytes
+        """
+        self.chain_handler = chain_handler
+        self.primary_addr = primary_addr
+        self.init_vmc_attributes()
+
+    def init_vmc_attributes(self):
+        self._vmc_addr = get_valmgr_addr()
+        self._vmc_sender_addr = get_valmgr_sender_addr()
+        self._vmc_bytecode = get_valmgr_bytecode()
+        self._vmc_code = get_valmgr_code()
+        self._vmc_abi = get_valmgr_abi()
+        self.logger.debug("vmc_addr=%s", self._vmc_addr)
+        self.logger.debug("vmc_sender_addr=%s", self._vmc_sender_addr)
+
+    # vmc utils ####################################
+
+    def call_vmc(self,
+                 func_name,
+                 args,
+                 sender_addr=None,
+                 value=0,
+                 gas=TX_GAS,
+                 gas_price=GASPRICE):
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        tx_obj = mk_vmc_tx_obj(func_name, args, sender_addr, value, gas, gas_price)
+        result = self.chain_handler.call(tx_obj)
+        decoded_result = decode_vmc_call_result(func_name, result)
+        self.logger.debug(
+            "call_vmc: func_name=%s, args=%s, result=%s",
+            func_name,
+            args,
+            decoded_result,
+        )
+        return decoded_result
+
+    def send_vmc_tx(self,
+                    func_name,
+                    args,
+                    sender_addr=None,
+                    value=0,
+                    gas=TX_GAS,
+                    gas_price=GASPRICE):
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        tx_obj = mk_vmc_tx_obj(func_name, args, sender_addr, value, gas, gas_price)
+        tx_hash = self.chain_handler.send_transaction(tx_obj)
+        self.logger.debug(
+            "send_vmc_tx: func_name=%s, args=%s, tx_hash=%s",
+            func_name,
+            args,
+            tx_hash,
+        )
+        return tx_hash
+
+    # vmc related #############################
+
+    def sample(self, shard_id, sender_addr=None):
+        """sample(shard_id: num) -> address
+        """
+        if sender_addr is None:
+            sender_addr = self.primary_addr
+        address_in_hex = self.call_vmc('sample', [shard_id], sender_addr=sender_addr)
+        # TODO: should see if there is a better way to automatically change the address result from
+        #       hex to bytes in. Maybe in `decode_contract_call_result`?
+        return to_canonical_address(address_in_hex)
```

Comment:
Given that we *know* that `address_in_hex` is well formed we can instead just use `evm.utils.hexadecimal.decode_hex` to convert to bytes.

Reply:
Have done it:) Thank you.